tx-pool/src/error.rs

```diff
@@ -1,69 +1,68 @@
 //! The error type for Tx-pool operations
 
-use ckb_error::{Error, ErrorKind};
+use ckb_error::{
+    impl_error_conversion_with_adaptor, impl_error_conversion_with_kind, prelude::*, Error,
+    ErrorKind, InternalError, InternalErrorKind, OtherError,
+};
 use ckb_fee_estimator::FeeRate;
 use ckb_types::packed::Byte32;
-use failure::Fail;
-use tokio::sync::mpsc::error::TrySendError as TokioTrySendError;
+use tokio::sync::{mpsc::error::TrySendError, oneshot::error::RecvError};
 
 /// TX reject message
-#[derive(Debug, PartialEq, Clone, Eq, Fail)]
+#[derive(Error, Debug, PartialEq, Clone, Eq)]
 pub enum Reject {
     /// Transaction fee lower than config
-    #[fail(
-        display = "The min fee rate is {} shannons/KB, so the transaction fee should be {} shannons at least, but only got {}",
-        _0, _1, _2
-    )]
+    #[error("The min fee rate is {0} shannons/KB, so the transaction fee should be {1} shannons at least, but only got {2}")]
     LowFeeRate(FeeRate, u64, u64),
 
     /// Transaction exceeded maximum ancestors count limit
-    #[fail(display = "Transaction exceeded maximum ancestors count limit, try send it later")]
+    #[error("Transaction exceeded maximum ancestors count limit, try send it later")]
     ExceededMaximumAncestorsCount,
 
     /// Transaction pool exceeded maximum size or cycles limit,
-    #[fail(
-        display = "Transaction pool exceeded maximum {} limit({}), try send it later",
-        _0, _1
-    )]
+    #[error("Transaction pool exceeded maximum {0} limit({1}), try send it later")]
     Full(String, u64),
 
     /// Transaction already exist in transaction_pool
-    #[fail(display = "Transaction({}) already exist in transaction_pool", _0)]
+    #[error("Transaction({0}) already exist in transaction_pool")]
     Duplicated(Byte32),
 
     /// Malformed transaction
-    #[fail(display = "Malformed {} transaction", _0)]
+    #[error("Malformed {0} transaction")]
     Malformed(String),
 }
 
-impl From<Reject> for Error {
-    fn from(error: Reject) -> Self {
-        error.context(ErrorKind::SubmitTransaction).into()
-    }
-}
+impl_error_conversion_with_kind!(Reject, ErrorKind::SubmitTransaction, Error);
```

Comment:
I strongly recommend that adding a brief description about what the macro does. This suggestion can be applied to all the macro invocations in this PR.

Example

```suggestion
// Implement conversion from `Reject` to `Error` with the error kind `SubmitTransaction`.
impl_error_conversion_with_kind!(Reject, ErrorKind::SubmitTransaction, Error);
```